File: runtime/perl/units/rt/i64_trunc_s.pl
# requires: rt/trap
# -2^63 is itself an exact NV and in range; the nearest NV below it already
# truncates out of range, so >= is the correct lower bound (ADR-55).
use POSIX ();
sub i64_trunc_s {
    my $x = $_[0];
    Rt::trap('invalid conversion to integer') if $x != $x;
    Rt::trap('integer overflow') unless $x >= -9223372036854775808.0 && $x < 9223372036854775808.0;
    return POSIX::trunc($x) & 0xFFFFFFFFFFFFFFFF;
}

File: runtime/java/units/rt/i64_trunc_s.java
// requires: rt/trap
// i64.trunc_f32_s / i64.trunc_f64_s: trapping signed truncation (see
// i32_trunc_s). The upper bound is `>= 2^63` since 2^63 is not representable as
// i64.
static long i64_trunc_s(double x) {
    if (Double.isNaN(x)) {
        trap("invalid conversion to integer");
    }
    if (Double.isInfinite(x)) {
        trap("integer overflow");
    }
    double t = x < 0 ? Math.ceil(x) : Math.floor(x);
    if (t < -9223372036854775808.0 || t >= 9223372036854775808.0) {
        trap("integer overflow");
    }
    return (long) t;
}
